browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,457 @@
+        ISOLATED_WORLD_ID_CHROME_INTERNAL);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::GetHiddenClassIdSelectorsOnUI(
+    content::GlobalFrameRoutingId frame_id,
+    const GURL& url,
+    std::unique_ptr<base::ListValue> selectors) {
+  if (!selectors || IsVettedSearchEngine(url.host())) {
+    return;
+  }
+  for (size_t i = 0; i < selectors->GetSize(); i++) {
+    base::ListValue* selectors_list = nullptr;
+    if (!selectors->GetList()[i].GetAsList(&selectors_list) ||
+        selectors_list->GetSize() == 0) {
+      continue;
+    }
+    std::string json_selectors;
+    if (!base::JSONWriter::Write(*selectors_list, &json_selectors) ||
+        json_selectors.empty()) {
+      json_selectors = "[]";
+    }
+    // Building a script for stylesheet modifications
+    std::string new_selectors_script =
+        base::StringPrintf(kSelectorsInjectScript, json_selectors.c_str());
+    if (selectors_list->GetSize() != 0) {
+      auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+      if (!frame_host)
+        return;
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(new_selectors_script), base::NullCallback(),
+          ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+  }
+
+  if (!enabled_1st_party_cf_filtering_) {
+    auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+    if (!frame_host)
+      return;
+    frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(*g_observing_script), base::NullCallback(),
+        ISOLATED_WORLD_ID_CHROME_INTERNAL);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::ProcessURL(
+    content::RenderFrameHost* render_frame_host,
+    const GURL& url,
+    const bool do_non_scriptlets) {
+  content::CosmeticFiltersCommunicationImpl::CreateInstance(render_frame_host,
+                                                            this);
+  if (!render_frame_host || !ShouldDoCosmeticFiltering(web_contents(), url)) {
+    return;
+  }
+  g_brave_browser_process->ad_block_service()
+      ->GetTaskRunner()
+      ->PostTaskAndReplyWithResult(
+          FROM_HERE,
+          base::BindOnce(GetUrlCosmeticResourcesOnTaskRunner, url.spec()),
+          base::BindOnce(
+              &BraveCosmeticResourcesTabHelper::GetUrlCosmeticResourcesOnUI,
+              weak_factory_.GetWeakPtr(),
+              content::GlobalFrameRoutingId(
+                  render_frame_host->GetProcess()->GetID(),
+                  render_frame_host->GetRoutingID()),
+              url.spec(), do_non_scriptlets));
+}
+
+void BraveCosmeticResourcesTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle || !navigation_handle->HasCommitted())
+    return;
+  ProcessURL(navigation_handle->GetRenderFrameHost(),
+             web_contents()->GetLastCommittedURL(),
+             navigation_handle->IsInMainFrame());
+}
+
+void BraveCosmeticResourcesTabHelper::ResourceLoadComplete(
```

Comment:
@antonok-edm the extension updated whenever the dom changed, right? I think this ResourceLoadComplete is acting as a kind of proxy for dom changes, but that doesn't seem very reliable to me and also we're refetching the resources even though the url hasn't changed

Reply:
actually the mutation observer is still in the script so I'm not sure why this is needed unless we're running the script too early which could certainly be the case if it's called from DidFinishNavigation. There's a race condition and the DOM might not be loaded by the time the script runs the first time

Reply:
`DocumentAvailableInMainFrame` could be used for the main frame, but I don't think there's an equivalent for subframes. I think what we need is a RenderFrameObserver like DOMDistiller that calls back to the browser process to get the rules

Reply:
The more I think about this the more I wish we could do it like we do right now with the extension: Load the JS script once on all pages at `document_start`, and then supply selector / scriptlet data when we have it at various navigation events.

As it stands right now in this PR on some sites the JS won't ever load, which I guess we have optimized for. But on the sites where there will be cosmetic filtering we're injecting the same script multiple times, potentially causing a lot of parsing performance overhead (not to mention added complexity and possible timing issues). In fact, can we at least just store a variable to remember that it has been injected for a frame, and then simply call data-providing functions from then on?

Reply:
@bridiver there are
```
  // This method is invoked when the document in the given frame finished
  // loading. At this point, scripts marked as defer were executed, and
  // content scripts marked "document_end" get injected into the frame.
  virtual void DOMContentLoaded(RenderFrameHost* render_frame_host) {}

  // This method is invoked when the load is done, i.e. the spinner of the tab
  // will stop spinning, and the onload event was dispatched.
  //
  // If the WebContents is displaying replacement content, e.g. network error
  // pages, DidFinishLoad is invoked for frames that were not sending
  // navigational events before. It is safe to ignore these events.
  virtual void DidFinishLoad(RenderFrameHost* render_frame_host,
                             const GURL& validated_url) {}
```

Reply:
The JS injection has been moved out from tab helper to renderer process